### PR TITLE
l4lb: fix inability to properly update service

### PR DESCRIPTION
**Pull request by oblazek**
When cilium-agent is restarted the restored backends are put into `restoredBackendHashes` set so that obsolete ones can be properly deleted only when k8s sync is done.

This is valid in k8s case, but when cilium-agent is running in standalone L4LB mode there is no way to do sync. There are only the ServiceAPI calls.
Therefore with the current code it is not possible to restart l4lb and remove specific backend from a service without previous service update call (to let l4lb know all the backends - simulate the sync). This should not be needed.

This commit fixes it in l4lb case.

```release-note
l4lb: fix inability to properly update service after agent restart 
```
related slack thread: https://cilium.slack.com/archives/C07BRCWT7GC/p1721913208811839

### CODE REVIEW

**Review by borkmann — CHANGES_REQUESTED** [review]
Hey @oblazek, is this a concrete fix? From the description it sounded more like a cleanup fwiw. It would be great if the commit desc can be more elaborate on the concrete issue. If there is no issue, I think it might be better to leave as is. If there is an issue, the test needs to be against `!option.Config.LoadBalancerExternalControlPlane` instead.

**Comment by oblazek** [comment]
> Hey @oblazek, is this a concrete fix? From the description it sounded more like a cleanup fwiw. It would be great if the commit desc can be more elaborate on the concrete issue. If there is no issue, I think it might be better to leave as is. If there is an issue, the test needs to be against `!option.Config.LoadBalancerExternalControlPlane` instead.

Oh I had no idea this option exists! Yeah this was a specific problem which we had in production. The steps to reproduce are as follows:

1) stop the L4LB (can be done in the ./test/l4lb/test.sh)
2) remove/destroy 1 backend
3) start the L4LB
4) send an service update with the 1 backend removed, but the remaining untouched
5) the "removed" backend will still be present in the map

**Comment by oblazek** [comment]
If you run 2 backends in the `./test/l4lb/test.sh`  it can be easily reproduced...

**Comment by borkmann** [comment]
Ok, makes sense, then lets change the condition to the one proposed for the correct fix. Thanks!

**Comment by borkmann** [comment]
> If you run 2 backends in the `./test/l4lb/test.sh` it can be easily reproduced...

If you have a chance, perhaps it would be good to extend the test suite for this corner case?

**Comment by oblazek** [comment]
> > If you run 2 backends in the `./test/l4lb/test.sh` it can be easily reproduced...
> 
> If you have a chance, perhaps it would be good to extend the test suite for this corner case?

totally I can do that :)

**Comment by mhofstetter** [comment]
:wave: Hey @oblazek and thanks for the fix!

Heads up that https://github.com/cilium/cilium/pull/34078 has been merged to `main` in the meantime. It removes the config property `option.Config.LoadBalancerExternalControlPlane` in favor of using `Clientset.IsEnabled()` directly.

To be able to read this information in the `Service(Manager)` we have to inject the k8s clientset via [Hive](https://docs.cilium.io/en/latest/contributing/development/hive/) when initializing the `Service(Manager)`.

Following patch should help - would be nice if you could rebase and apply/verify. Thanks, and sorry for the inconvenience.

```
diff --git a/pkg/service/cell.go b/pkg/service/cell.go
index be68449afc..243bc16a9b 100644
--- a/pkg/service/cell.go
+++ b/pkg/service/cell.go
@@ -7,6 +7,7 @@ import (
 	"github.com/cilium/hive/cell"
 
 	"github.com/cilium/cilium/pkg/datapath/types"
+	k8sClient "github.com/cilium/cilium/pkg/k8s/client"
 	monitorAgent "github.com/cilium/cilium/pkg/monitor/agent"
 )
 
@@ -30,6 +31,7 @@ type serviceManagerParams struct {
 	MonitorAgent monitorAgent.Agent
 
 	HealthCheckers []HealthChecker `group:"healthCheckers"`
+	Clientset      k8sClient.Clientset
 }
 
 func newServiceInternal(params serviceManagerParams) *Service {
@@ -40,5 +42,5 @@ func newServiceInternal(params serviceManagerParams) *Service {
 		}
 	}
 
-	return newService(params.MonitorAgent, params.Datapath.LBMap(), params.Datapath.NodeNeighbors(), enabledHealthCheckers)
+	return newService(params.MonitorAgent, params.Datapath.LBMap(), params.Datapath.NodeNeighbors(), enabledHealthCheckers, params.Clientset.IsEnabled())
 }
diff --git a/pkg/service/service.go b/pkg/service/service.go
index db2e0bdaee..3da71c73dc 100644
--- a/pkg/service/service.go
+++ b/pkg/service/service.go
@@ -278,11 +278,12 @@ type Service struct {
 
 	backendConnectionHandler sockets.SocketDestroyer
 
-	backendDiscovery datapathTypes.NodeNeighbors
+	backendDiscovery       datapathTypes.NodeNeighbors
+	k8sControlplaneEnabled bool
 }
 
 // newService creates a new instance of the service handler.
-func newService(monitorAgent monitorAgent.Agent, lbmap datapathTypes.LBMap, backendDiscoveryHandler datapathTypes.NodeNeighbors, healthCheckers []HealthChecker) *Service {
+func newService(monitorAgent monitorAgent.Agent, lbmap datapathTypes.LBMap, backendDiscoveryHandler datapathTypes.NodeNeighbors, healthCheckers []HealthChecker, k8sControlplaneEnabled bool) *Service {
 	var localHealthServer healthServer
 	if option.Config.EnableHealthCheckNodePort {
 		localHealthServer = healthserver.New()
@@ -300,6 +301,7 @@ func newService(monitorAgent monitorAgent.Agent, lbmap datapathTypes.LBMap, back
 		backendConnectionHandler: backendConnectionHandler{},
 		backendDiscovery:         backendDiscoveryHandler,
 		healthCheckers:           healthCheckers,
+		k8sControlplaneEnabled:   k8sControlplaneEnabled,
 	}
 	svc.lastUpdatedTs.Store(time.Now())
 
@@ -1832,8 +1834,8 @@ func (s *Service) restoreServicesLocked(svcBackendsById map[lb.BackendID]struct{
 			svcBackendsById[backend.ID] = struct{}{}
 		}
 
-		// There is no way to synchronize backends in standalone L4LB case, so don't block their removal
-		if len(newSVC.backendByHash) > 0 && !option.Config.LoadBalancerExternalControlPlane {
+		// There is no way to synchronize backends in standalone L4LB case with external control plane, so don't block their removal
+		if len(newSVC.backendByHash) > 0 && s.k8sControlplaneEnabled {
 			// Indicate that these backends were restored from BPF maps,
 			// so that they are not removed until SyncWithK8sFinished()
 			// is executed (if not observed in the meanwhile) to prevent
diff --git a/pkg/service/service_test.go b/pkg/service/service_test.go
index 1131f51ee8..491808fbae 100644
--- a/pkg/service/service_test.go
+++ b/pkg/service/service_test.go
@@ -220,7 +220,7 @@ func setupManagerTestSuite(tb testing.TB) *ManagerTestSuite {
 }
 
 func (m *ManagerTestSuite) newServiceMock(lbmap datapathTypes.LBMap) {
-	m.svc = newService(&FakeMonitorAgent{}, lbmap, nil, nil)
+	m.svc = newService(&FakeMonitorAgent{}, lbmap, nil, nil, true)
 	m.svc.backendConnectionHandler = testsockets.NewMockSockets(make([]*testsockets.MockSocket, 0))
 }
 
@@ -730,7 +730,7 @@ func TestRestoreServiceWithStaleBackends(t *testing.T) {
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
 			lbmap := mockmaps.NewLBMockMap()
-			svc := newService(&FakeMonitorAgent{}, lbmap, nil, nil)
+			svc := newService(&FakeMonitorAgent{}, lbmap, nil, nil, true)
 
 			_, id1, err := svc.upsertService(service("foo", "bar", "172.16.0.1", backendAddrs...))
 			require.NoError(t, err, "Failed to upsert service")
@@ -740,7 +740,7 @@ func TestRestoreServiceWithStaleBackends(t *testing.T) {
 			require.ElementsMatch(t, backendAddrs, toBackendAddrs(maps.Values(lbmap.BackendByID)), "lbmap not populated correctly")
 
 			// Recreate the Service structure, but keep the lbmap to restore services from
-			svc = newService(&FakeMonitorAgent{}, lbmap, nil, nil)
+			svc = newService(&FakeMonitorAgent{}, lbmap, nil, nil, true)
 			require.NoError(t, svc.RestoreServices(), "Failed to restore services")
 
 			// Simulate a set of service updates. Until synchronization completes, a given service
@@ -2261,7 +2261,7 @@ func TestRestoreServicesWithLeakedBackends(t *testing.T) {
 	m.svc.lbmap.AddBackend(backend5, backend5.L3n4Addr.IsIPv6())
 	require.Equal(t, len(backends)+4, len(m.lbmap.BackendByID))
 	lbmap := m.svc.lbmap.(*mockmaps.LBMockMap)
-	m.svc = newService(&FakeMonitorAgent{}, lbmap, nil, nil)
+	m.svc = newService(&FakeMonitorAgent{}, lbmap, nil, nil, true)
 
 	// Restore services from lbmap
 	err := m.svc.RestoreServices()
```

**Comment by oblazek** [comment]
will do that! thanks :)

**Comment by oblazek** [comment]
patch works like a charm @mhofstetter ;)

**Review by mhofstetter — APPROVED** [review]
LGTM, thanks @oblazek !

> > > If you run 2 backends in the `./test/l4lb/test.sh` it can be easily reproduced...
> > 
> > 
> > If you have a chance, perhaps it would be good to extend the test suite for this corner case?
> 
> totally I can do that :)

It would still be nice to have this covered :pray:

**Review by borkmann — APPROVED** [review]
ack, assuming also test case gets added :)

**Comment by oblazek** [comment]
Yeah, I wanted to refactor the l4lb test in a similar way to the nat46x64 (with functions). So will send that in a separate PR.
